### ISSUE TITLE
The RectAperture function with an off-axis distance from the optical axis as a parameter is not working    

### ISSUE THREAD

**Issue by Prerana2607** [opticspy/lightpipes#86]
When using the **RectAperture** function in LightPipes to simulate an off-axis beam through a thick aspheric lens, the resulting intensity pattern is unexpectedly symmetric. The expected behavior for an off-axis beam is to produce an asymmetric intensity pattern, whereas an on-axis beam should produce a symmetric pattern centered around the screen.

In the case of an off-axis beam, the beam enters at an angle relative to the optical axis, which would cause the intensity pattern to shift and become asymmetric. Additionally, the peak intensity should potentially decrease as part of the beam could be clipped by the aperture. However, in simulations using LightPipes, no such asymmetry is observed, and the intensity pattern appears similar to the on-axis case.

**Expected behaviour:**

1. Asymmetric intensity pattern, shifted in the direction of the offset for an off-axis beam
2. Potential reduction in peak intensity for the off-axis beam due to clipping by the aperture

Code Snippet to Reproduce the Issue:
```
from LightPipes import *
import numpy as np
import matplotlib.pyplot as plt

#all values in meters
wavelength = 450e-9 
focal_length = 3.1e-3
aperture_size = 200e-6 
off_axis_distance = 500e-6
propagation_distance=0.09
paraxial_angle=16.16 #in degrees
lens_thickness = 3.214e-3  # Lens center thickness in meter
n_lens=1.5964 #calculated refractive index of D-ZK3 @450 nm using linear interpolation method (source: https://www.edmundoptics.com/viewdocument/D-ZK3.pdf)
diameter=9.2e-3 #outer diameter of the lens in meters
CA1=3.84e-3 #clear aperture of surface 1 in meters
CA2=5e-3 #clear aperture of surface 2 in meters
R1=2.979042e-3 #radius of curvature of first surface in meters
c1=1/R1 #lens curvature of surface 1
R2= -2.725382e-3 #radius of curvature of second surface in meters
c2=1/R2 #lens curvature of surface 2
k1=-12.102353 #conic constant of surface 1
k2=-0.542123 #conic constant of surface 2
propagation_distance=0.09#in meter (90mm)
paraxial_angle=26.16 #in degrees

size=3e-3   #in meters
N=7000

asph_coeffs1 = [-8.068905e-3, 1.725894e-3, -1.839072e-4, 8.936267e-6]  # aspheric coefficients: a4, a6, a8, a10

# Aspheric coefficients for the second surface
asph_coeffs2 = [2.077710e-5, 1.373930e-4, -2.029328e-5, 2.961194e-6]  # aspheric coefficients: a4, a6, a8, a10


#User defined function for the aspheric thick lens:
def ThickLens(Fin, asph_coeffs1, asph_coeffs2, thickness, n_lens, diameter, CA1, CA2, c1,c2, k1, k2, x_shift=off_axis_distance, y_shift=0.0):
    """Simulate a thick aspheric lens with variable curvature across its surface.
    Parameters:
        Fin (Field): Input field.
        asph_coeffs1 (list): Aspheric coefficients for the first surface [c, k, A4, A6, ...].
        asph_coeffs2 (list): Aspheric coefficients for the second surface [c, k, A4, A6, ...].
        thickness (float): Center thickness of the lens.
        n_lens (float): Refractive index of the lens.
        diameter (float): Diameter of the lens.
        CA1 (float): Clear aperture diameter of the first surface.
        CA2 (float): Clear aperture diameter of the second surface.
        x_shift (float): Horizontal shift from the optical axis.
        y_shift (float): Vertical shift from the optical axis.
    Returns:
        Field: The output field after passing through the lens.
    """
    def aspheric_surface(F, asph_coeffs, CA,lens_curvature, conic_constant):#input parameters are the input field object, aspheric coefficients, and clear apertures (diameter of the lens through which light can pass)
       
        yy, xx = F.mgrid_cartesian# mgrid_cartesian refers to a method that generates two 2D arrays, xx and yy. These arrays represent the x and y coordinates of a Cartesian plane. Each point on this plane corresponds to a point in the optical field F.
        # a mesh grid of Cartesian coordinates that correspond to the field of light being simulated. This mesh grid is used to evaluate functions over a grid spanning the entire simulation area,
        xx -= x_shift
        yy -= y_shift
        r_squared = xx**2 + yy**2#compute the radial distance (r) from the center of the lens (or optical axis), adjusted for any lateral shifts
        r = np.sqrt(r_squared)
        
        # Calculate the aspheric surface profile
        z = lens_curvature * r_squared / (1 + np.sqrt(1 - (1 + conic_constant) * lens_curvature**2 * r_squared))#surface profile equation (https://www.edmundoptics.es/knowledge-center/application-notes/optics/an-in-depth-look-at-spherical-aberration-compensation-plates/)
        for i, coeff in enumerate(asph_coeffs):
            z += coeff * r**(2 * i + 4)#z+(Aspheric coefficients*(radial distance from the optical axis**4;4 when i=0)) + (Aspheric coefficients*(radial distance from the optical axis**6;6 when i=1))+(Aspheric coefficients*(radial distance from the optical axis**8;8 when i=2))
        
        # Apply the aperture mask for the clear aperture
        mask = r_squared <= (CA / 2)**2 #mask is a boolean value and CA/2 is the radius of the clear aperture. 
        #The radius value is squared so as to make it mathematically efficient and valid.
        #we consider the radius of the clear aperture in the above step instead of the diameter because
        #optical calculations often revolve around radial distances from the center of the lens,
        phase_shift = -2 * np.pi / F.lam * (n_lens - 1) * z #-2 * np.pi / wavelength * (index_difference) * optical_path_difference computes 
        #the additional phase (in radians)
        #There is a phase delay, thus the negative sign in the above step. The phase delay is because of the material present. 
        #Since the wave takes longer to traverse the material than it would take to travel the same geometrical distance in a vacuum.
        #The n-1  factor arises because the phase shift is only due to the extra path length induced by the presence of the material 
        #compared to air. F.lam is the wavelength. Wavelength parameter is not used so as to ensure that the wavelength used in calculations is the 
        #exact wavelength associated with the Field object. 
        phase_shift[~mask] = 0  # Ensures that the phase shift is only applied to points within the clear aperture. Phase shift outside the clear aperture
        #is 0. Thus there wont be any optical effect outside the aperture area.
        F.field[mask] *= np.exp(1j * phase_shift[mask])#applies the calculated phase shifts to the light field within the clear aperture.
        #F.field[mask] selects the portion of the light field within the clear aperture.
        #np.exp(1j * phase_shift[mask]) computes the complex exponential of the phase shifts, which represents the transformation of the wavefront 
        #due to the lens. 1j indicates that the phase shift is a complex exponential

        return F
   

    # Apply aperture to limit the beam to the lens diameter
    F = CircAperture(Fin, diameter/2) # Apply aperture to limit the beam to the lens diameter
    F = aspheric_surface(F, asph_coeffs1, CA1,c1, k1)# First spherical surface
    F = Forvard(F, thickness)# Propagate through the thickness of the lens
    F = aspheric_surface(F, asph_coeffs2, CA2,c2, k2)# Second spherical surface

    return F



F = Begin(size, wavelength, N)
F = RectAperture(F, aperture_size, aperture_size,x_shift=off_axis_distance,y_shift=0,angle=np.radians(paraxial_angle))
F=ThickLens(F, asph_coeffs1, asph_coeffs2, lens_thickness, n_lens, diameter, CA1, CA2, c1, c2, k1, k2, x_shift=off_axis_distance, y_shift=0.0)
F= Forvard(F, propagation_distance)
I = Intensity(F)

plt.imshow(I, extent=[-size/2*1e3, size/2*1e3, -size/2*1e3, size/2*1e3])
plt.title(f'No saturation Intensity Pattern Using Forvard Propagator\nGrid Size: {size*1e3:.2f} mm, Grid Points: {N}')
plt.xlabel('x (mm)')
plt.ylabel('y (mm)')
plt.colorbar(label='Intensity')
plt.show()
```
Here are the 2 plots for both the cases
1. without shift ![D90(mag2 5)_A200_simulation](https://github.com/user-attachments/assets/032877b9-939c-41ea-bf7f-b892a98de516)
2. With a shift of 500um in the x direction. ![D90(mag2 5)_A200_X+500um_simulation](https://github.com/user-attachments/assets/470824f9-ca83-413e-863b-e34dade430ad)

Below are the experimental data of same aperture size and offaxis shift distance 
1. with a shift![D90mm_A200um_C330TMD_SI4650um_X+500um_No-Saturation](https://github.com/user-attachments/assets/c2b0b032-5ba7-4dab-9107-a51bbd264d28)
2. without a shift ![D90mm_A200um_C330TMD_SI4650um_No-Saturation](https://github.com/user-attachments/assets/ca2e0143-055b-45a2-b3f6-1bcd8e19a13c)

**Comment by Prerana2607** [comment]
There was a problem with this code . Therefore I am taking down the issue.

**Comment by FredvanGoor** [comment]
Hi Prerana,Ok. I was wondering what was wrong with this very simple command....Please go on with your thick lens command, interesting!Best regards,Fred van GoorVerstuurd vanaf mijn iPadOp 18 sep 2024 om 01:46 heeft Prerana Chakrabarti ***@***.***> het volgende geschreven:﻿
Closed #86 as completed.

**Comment by 8bit-Dude** [comment]
Dear Fred,
I am working with Prerana on this problem. We are trying to simulate a Gaussian beam going through an off-axis aperture and then through a thick lens.
We thought that we could use the ABCD code (which would have been super-convenient), but it is only compatible with "pure" Gaussian beams.
Using the current code, is there any way we can propagate a beam through a small aperture and then a thick lens?
If not, how difficult is it to adapt the ABCD code for non-pure Gaussian beams?
Thanks in advanced for the reply!
